Add authentication page

sgomez opened the pull request:
Closes #14

sgomez committed:
Add Surfnet GSSP bundle

sgomez committed:
Add encore bundle

sgomez committed:
Add base template

- Enable sass encore module

sgomez committed:
Add ReadySession Serializer

- ReadySession serializer is required to save it into the user session
- Fix bug in ReadySessionId::equals

sgomez committed:
Saves ReadySession in ValueStore

Service to save and recover ReadySession into user browser session

sgomez committed:
Resolve ReceivedAuthnRequest in controllers 

Read the HTTP request and returns a ReceivedAuthnRequest instance 
to controllers as a parameter.

sgomez committed:
Disable checkMissingIterableValueType error

Many issues because Symfony returns generic iterators.

sgomez committed:
Start authentication request flow

- Save the authentication requests in HTTP session
- Get a new ReadySession and saves it in HTTP session
- Redirects the user to the authentication controller

sgomez committed:
Add authorization entry point in encore

Packages added to encore:
- typescript
- react+redux
- react-bootstrap

sgomez committed:
Add authorization page

The controller gets the SAML AuthnRequest and the ReadySession
Checks the authorization was started and shows the QrCode to the user

MKodde commented on package.json:
```diff
@@ -0,0 +1,16 @@
+        "regenerator-runtime": "^0.13.2",
+        "webpack-notifier": "^1.6.0"
+    },
+    "license": "UNLICENSED",
```
Maybe set this to the Apache 2.0 license?

sgomez replied:
Sure, no problem.

MKodde commented on config/packages/webpack_encore.yaml:
```diff
@@ -0,0 +1,25 @@
+webpack_encore:
+    # The path where Encore is building the assets - i.e. Encore.setOutputPath()
+    output_path: '%kernel.project_dir%/public/build'
```
If you like, you can look at some of our recent GSSP projects for a peek at a sensible encore setup. At Ibuildings, we like to use Typescript for JS work (not sure that would be justifiable in this project). The [WebAuthn GSSP](https://github.com/OpenConext/Stepup-Webauthn) could be used for reference.

sgomez replied:
Thanks. I was reading that bundle. The main difference with my configuration is that I save the files inside the canonical '/assets' and not in '/public'.

MKodde commented on support/qa/phpstan.neon:
```diff
@@ -5,10 +5,4 @@ parameters:
-        -
-            message: '#Method [a-zA-Z0-9\\_]+::[a-zA-Z0-9]+\(\) return type has no value type specified in iterable type array#'
-            path: '**/Infrastructure/Serializer/*Normalizer.php'
+    checkMissingIterableValueType: false
```
Please consider adding the reason for disabling this test as a comment to this line.

MKodde commented on src/Infrastructure/Controller/SingleSignOnController.php:
```diff
@@ -20,17 +20,94 @@
 use Symfony\Component\Routing\Annotation\Route;
+use function sprintf;
 
+/**
+ * @Route("/saml/sso", name="readid_saml_sso", methods={"GET"})
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
 final class SingleSignOnController extends AbstractController
```
Please consider offloading some of the logic from this Controller to a service. This way you will probably decrease coupling in the controller. And this will make the code more easily testable.

You might want to do this at a later stage when all interactions are finalized and a more clear set of services can be extracted.

sgomez replied:
My bad. I forgot to remove this line. I moved the AuthnRequest extraction from the Http Request to an ArgumentResolver just to resolve that.

MKodde commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "compilerOptions": {
+    "target": "es5",
+    "lib": ["dom", "dom.iterable", "esnext"],
+    "allowJs": true,
+    "skipLibCheck": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "strict": true,
+    "forceConsistentCasingInFileNames": true,
+    "module": "esnext",
+    "moduleResolution": "node",
+    "resolveJsonModule": true,
+    "isolatedModules": true,
+    "rootDirs": ["assets"],
+    "baseUrl": "assets",
+    "jsx": "react"
+  },
+  "exclude": ["node_modules"],
+  "include": [
+    "./assets/**/*"
+  ]
+}
```
Looks like a sensible set of settings. At my company we use [these defaults](https://github.com/OpenConext/Stepup-Webauthn/blob/develop/tsconfig.json). Comparing the two, we have some more strict settings explicitly enabled. Other than that, your setup looks good to go!

MKodde replied:
Also, adding Jest to test the JS code might be a good idea, for now, the low complexity might not warrant this, but if complexity is increased this might become a good idea.

sgomez replied:
I use this because is the default with create-react-scripts. I have no problem to change it.

MKodde reviewed (commented):
Once again, very nice work @sgomez! Nice to see you adopt modern web dev standards. I've left some suggestions in the code review section below. I have not yet tested this in a browser yet. But planning to do so in the near future. Please do not let this stop you from moving forward!

Regarding the licensing, for SURFConext/OpenConext work we use the Apache 2.0 license. We can postpone setting license entries until we finished the project. But I think we should start thinking about what licencing form to use. Do you have objections against using Apache 2.0?

sgomez committed:
Change typescript configuration

- Moved assets from /assets/js to /assets/typescript
- Activate more strict type-cheking

sgomez committed:
Set package.json LICENSE

sgomez committed:
Add explanation to disable iterable type checking

sgomez committed:
Remove suppress warning annotation

This issue was resolved before commit the class and the annotation was
left by mistake

sgomez committed:
Add parameters in template config

sgomez committed:
Improve class name for consistency
